layers/+lang/c-c++/funcs.el

```diff
@@ -182,3 +185,167 @@ and the arguments for flyckeck-clang based on a project-specific text file."
+  "Setup LSP backend auto-completion."
+  (if (configuration-layer/layer-used-p 'lsp)
+    (progn
+      (spacemacs|add-company-backends :backends company-lsp :modes c-mode-common)
+      ;;Disable client-side cache and sorting, as server does a better job
+      (setq company-transformers nil company-lsp-async t company-lsp-cache-candidates nil))
+    (warning "`lsp' layer is not installed, please add `lsp' layer to your dotfile.")))
+
+(defun spacemacs//c-c++-lsp-wrap-functions ()
+  "Wrap navigation functions for the LSP backend specified by the `c-c++-backend' configuration variable."
+  (defun c-c++/call-hierarchy () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-call-hierarchy" nil))
+  (defun c-c++/call-hierarchy-inv () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-call-hierarchy" t))
+  (defun c-c++/inheritance-hierarchy () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-inheritance-hierarchy"))
+  (defun c-c++/inheritance-hierarchy-inv () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-inheritance-hierarchy" t))
+  (defun c-c++/member-hierarchy () (interactive) (spacemacs//c-c++-lsp-funcall-interactively-no-args nil "-member-hierarchy"))
+  (defun c-c++/freshen-index () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-freshen-index"))
+  (defun c-c++/preprocess-file () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-preprocess-file")))
+
+(defun spacemacs//c-c++-lsp-bind-keys-for-mode (mode)
+  "Bind LSP backend functions for the specificed mode."
+  (spacemacs/set-leader-keys-for-major-mode mode
+    ;; goto
+    ;; FIXME gf conflicts with `lsp-ui-flycheck-list' binding in lsp layer
+    ;; Better corrected in lsp-layer? rebinding here for now
+    "ge" #'lsp-ui-flycheck-list
+    "gf" 'find-file-at-point
+    "gF" 'ffap-other-window
+    "g&" #'c-c++/references-address
+    "gR" #'c-c++/references-read
+    "gW" #'c-c++/references-write
+    "gc" #'c-c++/callers
+    "gv" #'c-c++/vars
+    ;; help/hierarchy
+    "hb" #'c-c++/base ;;Replace this with lsp-goto-implementation in lsp-layer?
+    "hd" #'c-c++/derived
+    "hc" #'c-c++/call-hierarchy
+    "hC" #'c-c++/call-hierarchy-inv
+    "hi" #'c-c++/inheritance-hierarchy
+    "hI" #'c-c++/inheritance-hierarchy-inv
+    "hm" #'c-c++/member-hierarchy
+    ;; jump
+    ;; lsp/backend
+    "lf" #'c-c++/freshen-index
+    "lp" #'c-c++/preprocess-file))
+
+(defun spacemacs//c-c++-lsp-customise-lsp-ui-peek ()
+  "Wrap some backend-specific features using the `lsp-ui-peek' functions provided by the `lsp' layer (`lsp-ui' package)"
+  (spacemacs//c-c++-lsp-call-function "spacemacs//c-c++-lsp-" "-customise-lsp-ui-peek")
+
+  (defun c-c++/callers () (interactive) (lsp-ui-peek-find-custom 'callers (spacemacs//c-c++-lsp-string "$" "/callers")))
+  (defun c-c++/derived () (interactive) (lsp-ui-peek-find-custom 'derived (spacemacs//c-c++-lsp-string "$" "/derived")))
```

Comment:
`$ccls/derived` `$cquery/derived` have both been removed. Users should instead use lsp-goto-implementation

ccls has a flattened `$ccls/inheritanceHierarchy` to provide multi-level information but I think it is too implementation-specific and does not worth the common configuration. https://github.com/MaskRay/ccls/wiki/Emacs#ccls-cross-reference-extension
```elisp
(defun ccls/derived ()
  (interactive)
  (lsp-ui-peek-find-custom 'derived "$ccls/inheritanceHierarchy"
                           (append (lsp--text-document-position-params) '(:flat t :level 3 :derived t))))
```